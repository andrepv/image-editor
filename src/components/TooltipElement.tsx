import React, { useEffect, useRef } from "react";
import tooltipPosition from "../helpers/setTooltipPosition";

type Props = {
  content: string;
  placement: string;
  targetEl: HTMLDivElement | null;
}

const TooltipElement: React.FC<Props> = ({content, placement, targetEl}) => {
  const tooltipRef = useRef(null);
  const triangleRef = useRef(null);

  const setTooltipPosition = () => {
    tooltipPosition.setTooltipPosition(
      placement,
      targetEl,
      tooltipRef.current,
      triangleRef.current,
    );
  };
  useEffect(setTooltipPosition);

  return (
    <div
      className={`tooltip tooltip_${placement} hidden`}
      ref={tooltipRef}
    >
      <div
        className="tooltip__triangle"
        ref={triangleRef}
      ></div>
      {content}
    </div>
  );
};

export default TooltipElement;