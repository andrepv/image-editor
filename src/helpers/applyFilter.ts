import ImageFilters from "./imagefilters";
const pixelsJS = require("../../node_modules/pixels.js/dist/Pixels.js");

export default function applyFilter(
  imageData: any,
  filterName: string,
  options: any[],
): ImageData {
  const firstTypeFilters = [
    "BoxBlur",
    "ColorTransformFilter",
    "Dither",
    "Emboss",
    "Enrich",
    "HSLAdjustment",
    "BrightnessContrastGimp",
    "StackBlur",
    "Mosaic",
    "Oil",
    "Posterize",
    "Sepia",
    "Sharpen",
  ];
  if (firstTypeFilters.includes(filterName)) {
    if (!options.length) {
      return (ImageFilters as any)[filterName](imageData);
    }
    const optionValues = options.map((option: any) => option.value);
    return (ImageFilters as any)[filterName](imageData, ...optionValues);
  }
  return pixelsJS.filterImgData(imageData, filterName);
}