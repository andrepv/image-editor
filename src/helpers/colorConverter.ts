export function hexToRgb(hex: string): string {
  var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
  return result ? [
    parseInt(result[1], 16),
    parseInt(result[2], 16),
    parseInt(result[3], 16),
   ].join(",") : "";
}

export function rgbToHex(r: number, g: number, b: number): string {
  return "#" + ((1 << 24) + (r << 16) + (g << 8) + b).toString(16).slice(1);
}