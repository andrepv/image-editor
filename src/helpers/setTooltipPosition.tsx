import { getCoords } from "./getCoords";

interface IElementData {
  top: number;
  right: number;
  bottom: number;
  left: number;
  width: number;
  height: number;
}

interface ICoords {
  x: number;
  y: number;
}

const initialElementData: IElementData = {
  top: 0,
  right: 0,
  bottom: 0,
  left: 0,
  width: 0,
  height: 0,
};

export class TooltipPosition {
  private placement: string = "";
  private tooltipElement: HTMLDivElement | null = null;
  private triangleElement: HTMLDivElement | null = null;
  private targetEl: IElementData = {...initialElementData};
  private tooltip: IElementData = {...initialElementData};
  private bodyWidth: number = document.body.offsetWidth;
  private offset: number = 10;

  setTooltipPosition(
    placement: string,
    targetEl: HTMLDivElement | null,
    tooltip: HTMLDivElement | null,
    triangle: HTMLDivElement | null,
  ): void {
    if (!tooltip || !targetEl || !triangle) {
      return;
    }

    this.placement = placement;
    this.tooltipElement = tooltip;
    this.triangleElement = triangle;
    this.targetEl = this.getTargetElementData(targetEl);
    this.tooltip = this.getTooltipElementData(tooltip);

    const coords = this.getTooltipCoords();
    if (coords) {
      this.tooltipElement.style.top = `${coords.y}px`;
      this.tooltipElement.style.left = `${coords.x}px`;
      this.tooltipElement.classList.remove("hidden");
    }
  }

  private getTooltipCoords(): ICoords | undefined {
    let coords;
    switch (this.placement) {
      case "right":
      case "left":
        coords = this.getHorizontalTooltipCoords();
        break;
      case "bottom":
      case "top":
        coords = this.getVerticalTooltipCoords();
        break;
    }
    if (coords) {
      return {
        x: coords.x,
        y: coords.y,
      };
    }
  }

  private getHorizontalTooltipCoords(): ICoords {
    const yCoord = (
      this.targetEl.top + this.targetEl.height / 2 - this.tooltip.height / 2
    );
    const xCoord = this.placement === "right"
    ? this.getRightTooltipXCoord()
    : this.getLeftTooltipXCoord();
    return {y: yCoord, x: xCoord};
  }

  private getVerticalTooltipCoords(): ICoords {
    const yCoord = this.placement === "bottom"
      ? this.getBottomTooltipYCoord()
      : this.getTopTooltipYCoord();
    const xCoord = this.getVerticalTooltipXCoord();
    return {y: yCoord, x: xCoord};
  }

  private getRightTooltipXCoord(): number {
    const isTipInViewport = (
      this.targetEl.right + this.tooltip.width < this.bodyWidth
    );
    const coord = isTipInViewport ? this.tooltip.right : this.tooltip.left;
    if (!isTipInViewport) {
      this.replaceTooltipClass("tooltip_left", "tooltip_right");
    }
    return coord;
  }

  private getLeftTooltipXCoord(): number {
    const isTipInViewport = this.targetEl.left - this.tooltip.width > 0;
    const coord = isTipInViewport ? this.tooltip.left : this.tooltip.right;
    if (!isTipInViewport) {
      this.replaceTooltipClass("tooltip_right", "tooltip_left");
    }
    return coord;
  }

  private getBottomTooltipYCoord(): number {
    const isTipInViewport = (
      this.tooltip.bottom + this.tooltip.height < window.innerHeight
    );
    const coord = isTipInViewport ? this.tooltip.bottom : this.tooltip.top;
    if (!isTipInViewport) {
      this.replaceTooltipClass("tooltip_top", "tooltip_bottom");
    }
    return coord;
  }

  private getTopTooltipYCoord(): number {
    const isTipInViewport = this.tooltip.top > 0;
    const coord = isTipInViewport ? this.tooltip.top : this.tooltip.bottom;
    if (!isTipInViewport) {
      this.replaceTooltipClass("tooltip_bottom", "tooltip_top");
    }
    return coord;
  }

  private getVerticalTooltipXCoord(): number {
    const targetEl小enter = this.targetEl.left + this.targetEl.width / 2;
    const tooltipRightEdge = targetEl小enter + this.tooltip.width / 2;
    const tooltipLeftEdge = targetEl小enter - this.tooltip.width / 2;
    let coord = tooltipLeftEdge;

    if (tooltipRightEdge > this.bodyWidth) {
      coord = this.bodyWidth - this.tooltip.width - this.offset;
      this.setTrianglePosition(targetEl小enter - coord);
    } else if (tooltipLeftEdge < 0) {
      coord = this.offset;
      this.setTrianglePosition(this.targetEl.width / 2);
    }
    return coord;
  }

  private setTrianglePosition(coord: number): void {
    if (this.triangleElement) {
      this.triangleElement.style.left = `${coord}px`;
    }
  }

  private replaceTooltipClass(
    newClassName: string,
    oldClassName: string,
  ): void {
    if (this.tooltipElement) {
      this.tooltipElement.classList.remove(oldClassName);
      this.tooltipElement.classList.add(newClassName);
    }
  }

  private getTargetElementData(targetEl: HTMLDivElement): IElementData {
    return {
      top: getCoords(targetEl).top,
      right: getCoords(targetEl).right,
      bottom: getCoords(targetEl).bottom,
      left: getCoords(targetEl).left,
      width: targetEl.offsetWidth,
      height: targetEl.offsetHeight,
    };
  }

  private getTooltipElementData(tooltip: HTMLDivElement): IElementData {
    return {
      top: this.targetEl.top - tooltip.offsetHeight - this.offset,
      right: this.targetEl.right + this.offset,
      bottom: this.targetEl.bottom + this.offset,
      left: this.targetEl.left - tooltip.offsetWidth - this.offset,
      width: tooltip.offsetWidth,
      height: tooltip.offsetHeight,
    };
  }
}

export default new TooltipPosition();