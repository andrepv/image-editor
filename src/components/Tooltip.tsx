import React, { useState, useRef } from "react";
import ReactDOM from "react-dom";
import TooltipElement from "./TooltipElement";

type Props = {
  content: string;
  placement: string;
  children: React.ReactNode;
}

const Tooltip: React.FC<Props> = ({content, placement, children}) => {
  const targetElRef = useRef(null);
  const [isVisible, toggleVisibility] = useState(false);
  const [isRemoved, remove] = useState(false);

  const show = () => toggleVisibility(true);
  const hide = () => toggleVisibility(false);
  const mouseMoveHandler = (event: React.MouseEvent) => {
    const target = event.target as HTMLElement;
    const isCursorOnTarget = target.closest(".tooltip__container");
    if (isCursorOnTarget && !isVisible && !isRemoved) {
      show();
    }
  };
  const clickHandler = () => {
    if (isVisible) {
      hide();
      remove(true);
    }
  };

  return (
    <>
      <div
        className="tooltip__container"
        ref={targetElRef}
        onMouseEnter={show}
        onMouseLeave={hide}
        onMouseMove={mouseMoveHandler}
        onClick={clickHandler}
      >
        {children}
      </div>
      {isVisible && ReactDOM.createPortal(
        <TooltipElement
          content={content}
          placement={placement}
          targetEl={targetElRef.current}
        />,
        document.body,
      )}
    </>
  );
};

export default Tooltip;