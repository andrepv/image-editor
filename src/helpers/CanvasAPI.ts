import { fabric } from "fabric";
import Cropper from "./Cropper";

type CanvasSize = {
  width: number;
  height: number;
}

export default class CanvasAPI {
  public canvas: fabric.Canvas;
  public imageElement: HTMLImageElement;
  public canvasSize: CanvasSize = {width: 0, height: 0};
  private cropper: Cropper;
  private currentMode: string = "";

  constructor(canvas: fabric.Canvas) {
    this.canvas = canvas;
    this.imageElement = new Image();
    this.cropper = new Cropper(this);
  }

  public renderImage(imageUrl: string, scale: number, mode: string): void {
    if (!imageUrl) {
      return;
    }
    this.imageElement.src = imageUrl;
    fabric.Image.fromURL(imageUrl, () => {
      const { width, height } = this.getImageSize(scale);
      this.setCanvasSize(width, height);

      this.canvas.clear();
      const imgInstance = new fabric.Image(this.imageElement, {
        selectable: false,
        hoverCursor: "default",
      });
      imgInstance.scaleToWidth(width);
      imgInstance.scaleToHeight(height);
      this.canvas.add(imgInstance);

      if (!mode && this.currentMode) {
        this.destroyCurrentMode();
        return;
      }

      if (mode) {
        this.initializeMode(mode);
      }
    });
  }

  private destroyCurrentMode(): void {
    if (this.currentMode === "crop") {
      this.currentMode = "";
      this.cropper.destroy();
    }
  }

  private initializeMode(mode: string): void {
    if (mode === "crop") {
      this.cropper.initialize();
      this.currentMode = mode;
    }
  }

  private getImageSize(scale: number): {
    width: number;
    height: number;
  } {
    const containerHeight = 0.85 * window.innerHeight * scale;
    const ratio = this.imageElement.width / this.imageElement.height;
    const height = Math.min(containerHeight / ratio, containerHeight);
    const width = ratio * height;
    return { width, height };
  }

  private setCanvasSize(width: number, height: number): void {
    this.canvasSize = {width, height};
    this.canvas.setHeight(height);
    this.canvas.setWidth(width);
  }
}