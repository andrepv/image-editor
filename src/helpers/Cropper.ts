import CanvasAPI from "./CanvasAPI";
import CropZone from "./CropZone";
import { fabric } from "fabric";

interface IEventTransform {
  corner: string,
  original: fabric.Object,
  originX: string,
  originY: string,
  width: number,
  ex: number,
  ey: number,
  target: {
    width: number,
    height: number,
    top: number,
    left: number,
  }
}

export default class Cropper {
  private overlay: fabric.Object | null = null;
  private readonly canvasAPI: CanvasAPI;
  private readonly cropZone: CropZone;
  private readonly listeners: any;

  constructor(canvasAPI: CanvasAPI) {
    this.canvasAPI = canvasAPI;
    this.cropZone = new CropZone(canvasAPI);
    this.listeners = {
      onMouseDown: this.handleMouseDown.bind(this),
      onMouseUp: this.handleMouseUp.bind(this),
      onObjectMoving: this.handleObjectMoving.bind(this),
      onObjectScaling: this.handleObjectScaling.bind(this),
    };
  }

  public initialize(): void {
    this.addEventListeners();
    this.addOverlay();
    this.cropZone.initialize();
  }

  public destroy() {
    this.removeEventListeners();
    this.removeOverlay();
    this.cropZone.destroy();
  }

  private addEventListeners(): void {
    const { canvas } = this.canvasAPI;
    canvas.on("mouse:down", this.listeners.onMouseDown);
    canvas.on("mouse:up", this.listeners.onMouseUp);
    canvas.on("object:moving", this.listeners.onObjectMoving);
    canvas.on("object:scaling", this.listeners.onObjectScaling);
  }

  private removeEventListeners(): void {
    const { canvas } = this.canvasAPI;
    canvas.off("mouse:down", this.listeners.onMouseDown);
    canvas.off("mouse:up", this.listeners.onMouseUp);
    canvas.off("object:moving", this.listeners.onObjectMoving);
    canvas.off("object:scaling", this.listeners.onObjectScaling);
  }

  private handleMouseDown(event: fabric.IEvent): void {
    this.canvasAPI.canvas.selection = false;
    if (event?.target?.name === "cropzone") {
      this.cropZone.toggleFocus(true);
    }
  }

  private handleMouseUp(): void {
    this.cropZone.toggleFocus(false);
  }

  private handleObjectMoving(event: fabric.IEvent): void {
    if (event?.target?.name !== "cropzone") {
      return;
    }
    const {
      startX,
      startY,
      endX,
      endY,
      startLeft,
      startTop,
    } = this.getMovingEventData(event);

    const diffX = startX - endX;
    const diffY = startY - endY;

    const endLeft = startLeft - diffX;
    const endTop = startTop - diffY;

    this.cropZone.move(endLeft, endTop);
  }

  private handleObjectScaling(event: fabric.IEvent): void {
    if (event?.target?.name !=="cropzone") {
      return;
    }
    const {
      startLeft,
      startTop,
      endLeft,
      endTop,
      endX,
      endY,
      width,
      height,
      corner,
    } = this.getScalingEventData(event);

    const resizeData: any = {
      tl: {
        width: width - (endX - startLeft),
        height: height - (endY - startTop),
        x: endX,
        y: endY,
      },
      mt: {
        width: width,
        height: height - (endY - startTop),
        x: endLeft,
        y: endY,
      },
      tr: {
        width: width + (endX - (startLeft + width)),
        height: height - (endY - startTop),
        x: endLeft,
        y: endY,
      },

      mr: {
        width: width + (endX - (startLeft + width)),
        height: height,
        x: endLeft,
        y: endTop,
      },
      br: {
        width: width + (endX - (startLeft + width)),
        height: height + (endY - (startTop + height)),
        x: endLeft,
        y: endTop,
      },

      mb: {
        width: width,
        height: height + (endY - (startTop + height)),
        x: endLeft,
        y: endTop,
      },

      bl: {
        width: width - (endX - startLeft),
        height: height + (endY - (startTop + height)),
        x: endX,
        y: endTop,
      },
      ml: {
        width: width - (endX - startLeft),
        height: height,
        x: endX,
        y: endTop,
      },
    };
    this.cropZone.resize(resizeData[corner], corner);
  }

  private getScalingEventData(event: fabric.IEvent) {
    const transform = event.transform as IEventTransform;
    const {corner, target, original} = transform;
    const {x: endX, y: endY} = this.getScalingPointer(event.pointer);
    const {
      left: startLeft,
      top: startTop,
    } = original as {left: number, top: number};
    const {width, height, left: endLeft, top: endTop} = target;
    return {
      startLeft,
      startTop,
      endLeft,
      endTop,
      endX,
      endY,
      width,
      height,
      corner,
    };
  }

  private getMovingEventData(event: fabric.IEvent) {
    const transform = event.transform as IEventTransform;
    const {ex: startX, ey: startY} = transform;
    const {x: endX, y: endY} = event.pointer as fabric.Point;
    const {
      left: startLeft,
      top: startTop,
    } = transform.original as {left: number, top: number};
    return {startX, startY, endX, endY, startLeft, startTop};
  }

  private getScalingPointer(
    pointer: fabric.Point | undefined,
  ): {x: number, y: number} {
    if (!pointer) {
      return {x: 0, y: 0};
    }
    const {canvasSize} = this.canvasAPI;
    let {x, y} = pointer as fabric.Point;
    if (x < 0) {
      x = 0;
    }
    if (x > canvasSize.width) {
      x = canvasSize.width;
    }
    if (y < 0) {
      y = 0;
    }
    if (y > canvasSize.height) {
      y = canvasSize.height;
    }
    return {x, y};
  }

  private addOverlay(): void {
    this.overlay = new fabric.Rect({
      name: "overlay",
      fill: "rgba(0, 0, 0, 0.5)",
      selectable: false,
      hoverCursor: "default",
      width: this.canvasAPI.canvasSize.width,
      height: this.canvasAPI.canvasSize.height,
    });
    this.canvasAPI.canvas.add(this.overlay);
  }

  private removeOverlay(): void {
    if (this.overlay) {
      this.canvasAPI.canvas.remove(this.overlay);
    }
  }
}