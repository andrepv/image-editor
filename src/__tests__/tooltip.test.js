import React from "react";
import { shallow, mount } from "enzyme";
import Tooltip from "../components/Tooltip";
import TooltipElement from "../components/TooltipElement";
import { TooltipPosition } from "../helpers/setTooltipPosition";

describe("<Tooltip />", () => {
  const props = {
    content: "test",
    placement: "right",
  };

  it("should match the snapshot", () => {
    const component = shallow(
      <Tooltip {...props}>
        <div>Target Element</div>
      </Tooltip>
    );
    expect(component.html()).toMatchSnapshot();
  });

  it("should toggle <TooltipElement/> on hover", () => {
    const wrapper = mount(<Tooltip {...props} />);
    const tooltipClassName = `.tooltip_${props.placement}`;
    expect(wrapper.find(tooltipClassName)).toHaveLength(0);

    wrapper.find(".tooltip__container").simulate("mouseEnter");
    expect(wrapper.find(tooltipClassName)).toHaveLength(1);

    wrapper.find(".tooltip__container").simulate("mouseLeave");
    expect(wrapper.find(tooltipClassName)).toHaveLength(0);

    wrapper.find(".tooltip__container").simulate("mouseMove");
    expect(wrapper.find(tooltipClassName)).toHaveLength(1);
  });
});

describe("<TooltipElement />", () => {
  const props = {
    content: "test",
    placement: "right",
    targetEl: <div>Target Element</div>,
  };

  it("should match the snapshot", () => {
    const component = shallow(<TooltipElement {...props}/>);
    expect(component.html()).toMatchSnapshot();
  });

  it("setTooltipPosition method should be called", () => {
    TooltipPosition.prototype.setTooltipPosition = jest.fn();
    mount(<TooltipElement {...props}/>);
    expect(TooltipPosition.prototype.setTooltipPosition.mock.calls.length)
    .toBe(1);
  });
});

describe("TooltipPosition", () => {
  it("should be displayed according to the 'placement' parameter", () => {
    const tooltipPosition = new TooltipPosition();
    tooltipPosition.targetEl = {
      top: 100,
      right: 100,
      bottom: 100,
      left: 100,
      width: 50,
      height: 50,
    };
    tooltipPosition.tooltip = {
      top: 40,
      right: 110,
      bottom: 110,
      left: 10,
      width: 80,
      height: 50,
    };
    tooltipPosition.bodyWidth = 500;
    tooltipPosition.placement = "right";
    let coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", 110);
    expect(coords).toHaveProperty("y", 100);

    tooltipPosition.placement = "left";
    coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", 10);
    expect(coords).toHaveProperty("y", 100);

    tooltipPosition.placement = "top";
    coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", 85);
    expect(coords).toHaveProperty("y", 40);

    tooltipPosition.placement = "bottom";
    coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", 85);
    expect(coords).toHaveProperty("y", 110);
  });

  it("should change position if the width doesn't fit in the viewport", () => {
    const tooltipPosition = new TooltipPosition();
    tooltipPosition.targetEl = {
      top: 100,
      right: 100,
      bottom: 100,
      left: 90,
      width: 50,
      height: 50,
    };
    tooltipPosition.tooltip = {
      top: -10,
      right: 110,
      bottom: 110,
      left: -20,
      width: 100,
      height: 100,
    };
    tooltipPosition.bodyWidth = 100;
    tooltipPosition.replaceTooltipClass = jest.fn();
    tooltipPosition.setTrianglePosition = jest.fn();

    tooltipPosition.placement = "right";
    let coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", -20);
    expect(coords).toHaveProperty("y", 75);
    expect(tooltipPosition.replaceTooltipClass)
    .toBeCalledWith("tooltip_left", "tooltip_right");

    tooltipPosition.placement = "left";
    coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", 110);
    expect(coords).toHaveProperty("y", 75);
    expect(tooltipPosition.replaceTooltipClass)
    .toBeCalledWith("tooltip_right", "tooltip_left");

    tooltipPosition.placement = "top";
    coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", -10);
    expect(coords).toHaveProperty("y", 110);
    expect(tooltipPosition.replaceTooltipClass)
    .toBeCalledWith("tooltip_bottom", "tooltip_top");
    expect(tooltipPosition.setTrianglePosition).toBeCalledWith(125);

    tooltipPosition.placement = "bottom";
    tooltipPosition.offset = 10;
    window.innerHeight = 200;
    tooltipPosition.targetEl.left = 10;
    coords = tooltipPosition.getTooltipCoords();
    expect(coords).toHaveProperty("x", 10);
    expect(coords).toHaveProperty("y", -10);
    expect(tooltipPosition.replaceTooltipClass)
    .toBeCalledWith("tooltip_top", "tooltip_bottom");
    expect(tooltipPosition.setTrianglePosition).toBeCalledWith(25);
  });
});