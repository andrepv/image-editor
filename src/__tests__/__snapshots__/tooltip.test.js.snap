// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tooltip /> should match the snapshot 1`] = `"<div class=\\"tooltip__container\\"><div>Target Element</div></div>"`;

exports[`<TooltipElement /> should match the snapshot 1`] = `"<div class=\\"tooltip tooltip_right hidden\\"><div class=\\"tooltip__triangle\\"></div>test</div>"`;
